libsolutil/CDCL.h

```diff
@@ -48,16 +49,137 @@ struct Literal
 		return std::make_tuple(positive, variable) < std::make_tuple(_other.positive, _other.variable);
 	}
 };
-using Clause = std::vector<Literal>;
+
+struct Clause {
+	const Literal& front() const
+	{
+		return lits.front();
+	}
+	const Literal& back() const
+	{
+		return lits.back();
+	}
+	Literal& front()
+	{
+		return lits.front();
+	}
+	Literal& back()
+	{
+		return lits.back();
+	}
+	const Literal& operator[](const size_t at) const
+	{
+		return lits.at(at);
+	}
+	const Literal& at(const size_t at) const
+	{
+		return lits.at(at);
+	}
+	Literal& operator[](const size_t at)
+	{
+		return lits.at(at);
+	}
+	Literal& at(const size_t at)
+	{
+		return lits.at(at);
+	}
+	auto size() const
+	{
+		return lits.size();
+	}
+	auto empty() const
+	{
+		return lits.empty();
+	}
+	auto begin() {
+		return lits.begin();
+	}
+	auto end() {
+		return lits.end();
+	}
+	auto begin() const {
+		return lits.begin();
+	}
+	auto end() const {
+		return lits.end();
+	}
+	void push_back(const Literal& l)
+	{
+		lits.push_back(l);
+	}
+
+	std::vector<Literal> lits;
+	uint64_t ID;
+};
+
+class TriState {
+public:
+	TriState(const bool b)
+	{
+		if (b)
+			val = 1;
+		else
+			val = 0;
+	}
+
+	TriState()
+	{}
+
+	bool operator==(const TriState& other) const
+	{
+		return val == other.val;
+	}
+
+	bool operator!=(const TriState& other) const
+	{
+		return val != other.val;
+	}
+
```

Comment:
I was a bit hesitant, because the `~` on `tristate_unset` is unclear. I'd prefer if people had to use `!x.toBool()` which makes it clearer to the user that the underlying value can be more than just a boolean. What do you think? I mean, it is more verbose with `!x.toBool()`, and too much verbosity can also lead to bugs.

Reply:
Yeah, I'm also unsure on the semantics of negation, although negation of unset to be unset should be fine.

Reply:
We can add the negation operator, but it would make thing strange. Here is an example:

```
TriState a;
TriState b;
if (a == !b) cout << "Equal" << endl;
else cout << "Not equal" << end;
```

We get `Equal` output for this. However, if the comparison was: `a == b` (instead of `a == !b`), we would **still** get `Equal`... which I think is very strange and may make people start doubting their sanity :laughing:

Before introducing negation, I would rather wait for a good use-case where negation would help clean up the code, given the potential confusion it could cause. What do you think?